### PR TITLE
fix: 3.12 removed distutils, add fallback to setuptools

### PR DESCRIPTION
**Pull request by GabDug**
Hello!

`distutils` was removed in Python 3.12, burt is included in setuptools 48+.

We might want to add setuptools to the requirements for Python, maybe as extra?

I'm open for discussion for better ways to add 3.12 support :) 

Have a nice day!

### CODE REVIEW

File: tree_sitter/__init__.py [hunk 1]
```diff
@@ -2,8 +2,13 @@
...
 import enum
 from ctypes import c_void_p, cdll
-from distutils.ccompiler import new_compiler
-from distutils.unixccompiler import UnixCCompiler
+try:
+    from distutils.ccompiler import new_compiler
+    from distutils.unixccompiler import UnixCCompiler
+except ImportError:
+    from setuptools._distutils.ccompiler import new_compiler
+    from setuptools._distutils.unixccompiler import UnixCCompiler
```
Comment (amaanq) [inline tree_sitter/__init__.py:10]: I'm not a huge fan of this method, is there one import supported by 3.7-3.12? I'm not too familiar w/ what changed in 3.12

Comment (narpfel) [reply tree_sitter/__init__.py:10]: `setuptools>=60.0.0` includes a vendored copy of `distutils` ([ref](https://setuptools.pypa.io/en/latest/deprecated/distutils-legacy.html)), so the only change necessary is to depend on `setuptools` for Python >= 3.12. When `setuptools` is installed, importing `distutils` will automatically use the vendored copy from `setuptools`:

```pycon
>>> import distutils.unixccompiler
>>> distutils.unixccompiler
<module 'distutils.unixccompiler' from '/usr/lib/python3.11/site-packages/setuptools/_distutils/unixccompiler.py'>
```

Comment (amaanq) [reply tree_sitter/__init__.py:10]: I'm fine adding setuptools as a dependency, unless there's downsides to that

Comment (GabDug) [reply tree_sitter/__init__.py:10]: I'm updating the PR then. Setting the dependency as `extra` is another discussion I suppose :)

**Review by amaanq — APPROVED** [review]

**Comment by amaanq** [comment]
looks good, a bit annoying that Python removed distutils though..

**Comment by GabDug** [comment]
Closes #162 (by keeping distutils for the moment)

**Review by Carreau — APPROVED** [review]

**Comment by atapin** [comment]
@GabDug should 3.12 be added to the build matrix in the CI pipeline?

**Comment by GabDug** [comment]
Updated `cibuildwheel` to the latest version -- it adds wheels by default for CPython 3.12 and Pypy 3.10.

Not sure if it's needed in `ci.yml`: I see Pypy 3.9 and CPython 3.11 are not there, even though we build for these versions.

**Comment by GabDug** [comment]
Both workflows are running fine on my fork:
https://github.com/GabDug/py-tree-sitter/actions/runs/6721090411
https://github.com/GabDug/py-tree-sitter/actions/runs/6721090420 (release expected to fail on a fork)

**Comment by amaanq** [comment]
@GabDug this fails with 3.12 which I just added - do we need to install dependencies somehow first? How would that be done

**Comment by amaanq** [comment]
Thanks for the PR and your patience!

**Pull request merged**